src/OrchardCore/OrchardCore/Shell/Builders/ShellContainerFactory.cs

```diff
@@ -128,6 +128,17 @@ public async Task<IServiceProvider> CreateContainerAsync(ShellSettings settings,
             // If the startup is not coming from an extension, associate it to the application feature.
             // For instance when Startup classes are registered with Configure<Startup>() from the application.
 
+            // Check RequireFeaturesAttribute: only call ConfigureServices if all required features are enabled
+            var requiredFeatures = RequireFeaturesAttribute.GetRequiredFeatureNamesForType(startup.GetType());
```

Comment:
This is the fix for the problem. The other changes came as part of the investigating. But all the changes in this PR as good to keep./

Reply:
My first impression is that this is not correct. The startups should not even be in the DI if the requirements are not met. The check here should do that: https://github.com/OrchardCMS/OrchardCore/blob/a7734ff31255ea854be47c5fed575de49216497c/src/OrchardCore/OrchardCore/Shell/Builders/CompositionStrategy.cs#L43-L45

I'll need more time to check this further.